### PR TITLE
firstpullrquest

### PR DESCRIPTION
**Pull request by poetlifeart**
First review

### CODE REVIEW

File: dotproduct.py [hunk 1]
```diff
@@ -1,33 +1,32 @@
...
 
-class ScaleDotProductAttention(nn.Module):
-    def __init__(self):
-        super().__init__()
+class ScaleDotProductAttention(nn.Module):#
+    def __init__(self):#
+        super().__init__()#
 
-    def forward(self, k,q, v, mask=None):
+    def forward(self, k,q, v, mask=None):#
```
Comment (benglard) [inline dotproduct.py]: write as `q, k, v`. needs proper spacing of variable names

File: dotproduct.py [hunk 2]
```diff
@@ -1,33 +1,32 @@
...
 
-        self.softmax=nn.Softmax(dim=-1)
-        d_tensor=k.shape[-1]
-        k_t=k.transpose(2,3)
-
-        score=(q @ k_t)/math.sqrt(d_tensor)
+        self.softmax=nn.Softmax(dim=-1)#
```
Comment (benglard) [inline dotproduct.py]: no need for this object. Can just use softmax as a function below

File: dotproduct.py [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
+        d_tensor=k.shape[-1]#
+        k_t=k.transpose(2,3)#
 
+        score=(q @ k_t)/math.sqrt(d_tensor)#
```
Comment (benglard) [inline dotproduct.py]: just calculate the scale once in the constructor

File: dotproduct.py [hunk 4]
```diff
@@ -1,33 +1,32 @@
...
         if mask !=None :
-            score=score.masked_fill(mask==0, -1e9)
+            score=score.masked_fill(mask==0, -1e9)#
```
Comment (benglard) [inline dotproduct.py]: can use `float('-inf')` instead of -1e9

File: dotproduct.py [hunk 5]
```diff
@@ -1,33 +1,32 @@
...
 
-        score=self.softmax(score)
-        v=score @ v
+        score=self.softmax(score)#
```
Comment (benglard) [inline dotproduct.py]: var name should be clearer, like `att` or `attention` or `attention_matrix`

File: dotproduct.py [hunk 6]
```diff
@@ -1,33 +1,32 @@
...
+        v=score @ v#
```
Comment (benglard) [inline dotproduct.py]: needs proper spacing of `=` throughout

File: dotproduct.py (hunk 1 above)
Comment (benglard) [reply dotproduct.py]: also should add documentation of expectations of arguments, e.g. tensor shapes of q/k/v/mask

File: dotproduct.py [hunk 7]
```diff
@@ -1,33 +1,32 @@
...
-        return v, score
+        return v, score#
 
-k=torch.rand(2, 8, 3, 4)
-q=torch.rand(2, 8, 3, 4)
-v=torch.rand(2, 8, 3, 4)
-test=ScaleDotProductAttention()
-t, m=test(k,q,v)
-print(t.shape)
+k=torch.rand(2, 8, 3, 4)#
```
Comment (benglard) [inline dotproduct.py]: you can put these in a `if __name__ == '__main__'` block so this code is not executed every time

Comment (benglard) [reply dotproduct.py]: perhaps also name the dimensions for clarity like:
```python
batch = 2
timesteps = 8
heads = 3
feats = 4
q = torch.rand(batch, timesteps, heads, feats)
...
```

File: dotproduct.py (hunk 5 above)
Comment (benglard) [reply dotproduct.py]: can be something like `score.softmax(dim=-1)`

File: dotproduct.py (hunk 7 above)
Comment (poetlifeart) [reply dotproduct.py]: Got it  (note As transposition is on dimension 2 and 3, timesteps, feats have to be the last dimensions in that order.)

File: dotproduct.py (hunk 5 above)
Comment (poetlifeart) [reply dotproduct.py]: done

File: dotproduct.py [hunk 8]
```diff
@@ -1,33 +1,60 @@
-import math
+import math
```
Comment (benglard) [inline dotproduct.py]: extra space

File: dotproduct.py [hunk 9]
```diff
@@ -1,33 +1,60 @@
...
 import torch
 import torch.nn as nn
 
+#This class accepts queries, keys, and values as inputs and returns the output of the attention mechanism
+#as well as the attention weights. 
+
 class ScaleDotProductAttention(nn.Module):
-    def __init__(self):
+
+    # The constructor accepts the dimension of the key and value vectors as input.
```
Comment (benglard) [inline dotproduct.py]: look up `docstrings` for the right way to comment functions in python

File: dotproduct.py [hunk 10]
```diff
@@ -1,33 +1,60 @@
...
+    def __init__(self, d_tensor ):
         super().__init__()
 
-    def forward(self, k,q, v, mask=None):
+        #The square root of the dimension of the key feature vector is stored in the d_tensor attribute
```
Comment (benglard) [inline dotproduct.py]: please leave a space between `#` and the comment for readability throughout

File: dotproduct.py [hunk 11]
```diff
@@ -1,33 +1,60 @@
...
-        k_t=k.transpose(2,3)
+    # The forward method accepts queries, keys, and values as inputs and returns the output of the attention mechanism
+    # Each tensor has the following dimensions: batch size, number of heads, number of timesteps, and dimension of the feature vector
+    def forward(self, queries ,keys, values, mask=None):
 
-        score=(q @ k_t)/math.sqrt(d_tensor)
+        #by transposing the keys, we can compute the dot product between the queries and keys via matrix multiplication
+        #The queries and keys are multiplied together and divided by the square root of the dimension of the key vector
 
+        keys_transpose=keys.transpose(2,3)
```
Comment (benglard) [inline dotproduct.py]: readability

File: dotproduct.py [hunk 12]
```diff
@@ -1,33 +1,60 @@
...
 class ScaleDotProductAttention(nn.Module):
-    def __init__(self):
+
+    # The constructor accepts the dimension of the key and value vectors as input.
+    def __init__(self, d_tensor ):
         super().__init__()
 
-    def forward(self, k,q, v, mask=None):
+        #The square root of the dimension of the key feature vector is stored in the d_tensor attribute
+        self.d_tensor = math.sqrt(d_tensor)
```
Comment (benglard) [inline dotproduct.py]: should name this `self.scale` for clarity

File: dotproduct.py [hunk 13]
```diff
@@ -1,33 +1,60 @@
...
+        #by transposing the keys, we can compute the dot product between the queries and keys via matrix multiplication
+        #The queries and keys are multiplied together and divided by the square root of the dimension of the key vector
 
+        keys_transpose=keys.transpose(2,3)
+        attention = (queries @ keys_transpose)/self.d_tensor
+
+        #The mask is used to make the attention mechanism causal
         if mask !=None :
-            score=score.masked_fill(mask==0, -1e9)
+            attention = attention.masked_fill(mask == 0, '-inf')
```
Comment (benglard) [inline dotproduct.py]: bug: you mean `float('-inf')`

File: dotproduct.py [hunk 14]
```diff
@@ -1,33 +1,60 @@
...
 
-        score=(q @ k_t)/math.sqrt(d_tensor)
+        #by transposing the keys, we can compute the dot product between the queries and keys via matrix multiplication
+        #The queries and keys are multiplied together and divided by the square root of the dimension of the key vector
 
+        keys_transpose=keys.transpose(2,3)
+        attention = (queries @ keys_transpose)/self.d_tensor
+
+        #The mask is used to make the attention mechanism causal
         if mask !=None :
```
Comment (benglard) [inline dotproduct.py]: can you change this to `if mask is not None`

File: dotproduct.py [hunk 15]
```diff
@@ -1,33 +1,60 @@
...
-            score=score.masked_fill(mask==0, -1e9)
+            attention = attention.masked_fill(mask == 0, '-inf')
+
+        #The softmax function is applied to the attention weights to obtain the output of the attention mechanism
+        #The output of the attention mechanism is multiplied by the values to obtain the final output of the attention mechanism
+        attention = torch.softmax(attention, dim=-1)
+        values = attention @ values#
```
Comment (benglard) [inline dotproduct.py]: remove `#`

File: dotproduct.py [hunk 16]
```diff
@@ -1,33 +1,60 @@
...
+
+        return values, attention#
+
+#Testing the ScaleDotProductAttention class
+if __name__ == "__main__":
```
Comment (benglard) [inline dotproduct.py]: I would recommend you stick to either single quote or double quote for strings throughout. I tend to prefer single quotes

File: dotproduct.py [hunk 17]
```diff
@@ -1,33 +1,60 @@
...
+    #The queries, keys, and values are 4D tensors with the following dimensions: batch size, number of heads, number of timesteps, and dimension features
+    batch_size = 2
+    heads = 3
+    timesteps = 8
+    feats = 4
 
-        score=self.softmax(score)
-        v=score @ v
+    keys = torch.rand(batch_size, heads, timesteps, feats)
+    quereis = torch.rand(batch_size, heads, timesteps, feats)
```
Comment (benglard) [inline dotproduct.py]: misspelling

File: ecn.py [hunk 18]
```diff
@@ -1,37 +1,53 @@
...
-        pos=torch.arange(0, max_len, dtype=torch.float64)
-        _2i=torch.arange(0, d_model, step=2, dtype=torch.float64)
-       
-        arg=torch.outer(pos, (10000 ** (_2i / d_model)).pow_(-1))
+        # we use register_buffer to store the positional encoding as a buffer so that it is not updated during training
+        # furthermore, this allows for the positional encoding to be moved to the GPU, if available with the model parameters
+        self.register_buffer('encoding', torch.zeros((max_len, d_model), dtype=torch.float32))
+
+
+        # set up the a range of possible values for the positions of the words in the sequence
```
Comment (benglard) [inline ecn.py:21]: might as well write this as:
```
comment
line
comment
line
...
```
will be easier to read

File: ecn.py [hunk 19]
```diff
@@ -1,37 +1,53 @@
...
+        # setup the argument of the sinusoid function for the even and odd indices of the positins of the words
+        # note the argument is always given by an even number, even indecies map to sin and odd indices map to cos
```
Comment (benglard) [inline ecn.py]: spelling

File: ecn.py [hunk 20]
```diff
@@ -1,37 +1,53 @@
...
+        pos = torch.arange(0, max_len, dtype=torch.float32)
+        evenumbers = torch.arange(0, d_model, step=2, dtype=torch.float32)
+        arg = torch.outer(pos, (10000 **( -evenumbers / d_model)))
```
Comment (benglard) [inline ecn.py]: readability

File: ecn.py [hunk 21]
```diff
@@ -1,37 +1,53 @@
...
-position= PositionalEncoding (d_model=8, max_len=20)
-x=torch.arange(0, 20, dtype=torch.float64)
-x=x.unsqueeze(dim=0)
-print(position.forward(x).shape)
+# Testing the PositionalEncoding class
+# condition to run the code below
+if __name__ == "__main__":
+
+    # The constrcutor accepts the dimension of the feature vector of the word and the maximum length of the sequence as input
+    position = PositionalEncoding (d_model=8, max_len=40)
```
Comment (benglard) [inline ecn.py]: readability

File: ecn.py [hunk 22]
```diff
@@ -1,37 +1,53 @@
...
+    # The forward method accepts a batch of sequences as input and returns the positional encoding for the sequence
     def forward(self, x):
-        # self.encoding
-        # [max_len = 512, d_model = 512]
+        
+        # x is a 2D tensor with dimensions batch size and sequence length
+        # we use the sequence length to cutoff the positional encoding to the correct length
         batch_size, seq_len = x.shape
-        # [batch_size = 128, seq_len = 30]
         return self.encoding[:seq_len, :]
```
Comment (benglard) [inline ecn.py]: since the only shape parameter that matters here is `seq_len`, why restrict it to 2D tensors?

File: ecn.py [hunk 23]
```diff
@@ -1,37 +1,53 @@
...
+
+        # set up the a range of possible values for the positions of the words in the sequence
+        # setup the argument of the sinusoid function for the even and odd indices of the positins of the words
+        # note the argument is always given by an even number, even indecies map to sin and odd indices map to cos
+        pos = torch.arange(0, max_len, dtype=torch.float32)
+        evenumbers = torch.arange(0, d_model, step=2, dtype=torch.float32)
+        arg = torch.outer(pos, (10000 **( -evenumbers / d_model)))
+
         
+        # compute the positional encoding
```
Comment (benglard) [inline ecn.py:28]: not critical, but I think this would be cleaner and more readable if you do the computations first, and then do `self.register_buffer` last

File: ecn.py (hunk 22 above)
Comment (poetlifeart) [reply ecn.py]: I am not sure about this but let's discuss later. It is how the original code was.

**Comment by benglard** [comment]
LGTM
